### PR TITLE
Player Docs: Action Interrupt

### PR DESCRIPTION
**Pull request by fig02**
docs hopefully speak for themselves

Theres a bug caused by this system, known in the speedrunning community as "item stagger", which I will document at some point. Didn't want to do it in this pass though.

### CODE REVIEW

**Review by cadmic — APPROVED** [review]
out of curiosity, what's the TL;DR on what causes item stagger?

File: src/overlays/actors/ovl_player_actor/z_player.c [hunk 1]
```diff
@@ -3917,21 +3917,44 @@ s32 Player_TryActionChangeList(PlayState* play, Player* this, s8* actionChangeLi
     return false;
 }
 
-s32 func_808374A0(PlayState* play, Player* this, SkelAnime* skelAnime, f32 arg3) {
+typedef enum {
+    /* -1 */ PLAYER_INTERRUPT_NONE = -1,
+    /*  0 */ PLAYER_INTERRUPT_NEW_ACTION,
+    /*  1 */ PLAYER_INTERRUPT_MOVE
+} PlayerActionInterruptResult;
```
Comment (cadmic) [inline src/overlays/actors/ovl_player_actor/z_player.c:3924]: can you double-check retail bss ordering?

**Comment by fig02** [comment]
This isnt much of a tldr cause its hard to explain concisely.

Most usages of this system have the following usage:
```c
        interruptResult = Player_TryActionInterrupt(play, this, &this->skelAnime, 4.0f);

        if ((interruptResult != PLAYER_INTERRUPT_NEW_ACTION) &&
            ((interruptResult >= PLAYER_INTERRUPT_MOVE) || LinkAnimation_Update(play, &this->skelAnime))) {
```
Where the only way to exit the current action is to either have a "move" interrupt, or have the animation complete. 

However, the `updateUpperBody` argument passed to `Player_TryActionChangeList` (called by `Player_TryActionInterrupt`) is set to true. This means Player can use items during the interrupt window.
The problem with that is, some items can change players *main* animation to an idle looping animation, instead of a "play once" animation. `LinkAnimation_Update` will never return true to say a looping animation is "finished" cause it loops forever. 
So, link gets stuck in the current action hes in that tried to do the interrupt in the first place.

The cyclical "staggering" is because the upper body will only update during the configured interrupt window. So, for example when doing bonk item stagger with hookshot, the item pull animation will only happen in the last 5 frames of the looping idle animation, because thats when the interrupt window is happening.

**Comment by fig02** [comment]
I would say the bug here is that the upper body should not be allowed to update in the interrupt window. But they could have wanted using items to be a valid interrupt option. So I'm not really sure.

File: src/overlays/actors/ovl_player_actor/z_player.c (hunk 1 above)
Comment (fig02) [reply src/overlays/actors/ovl_player_actor/z_player.c:3924]: image

**Comment by cadmic** [comment]
interesting, thanks!

**Review by Dragorn421 — APPROVED** [review]

**Pull request merged**